File: code-recipes/java/src/main/java/vendorretailprocurement/shipments/SubmitShipmentsRecipe.java
package vendorretailprocurement.shipments;

import com.amazon.SellingPartnerAPIAA.LWAException;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.threeten.bp.OffsetDateTime;
import software.amazon.spapi.ApiException;
import software.amazon.spapi.api.vendor.shipments.v1.VendorShippingApi;
import software.amazon.spapi.models.vendor.shipments.v1.*;
import util.Constants;
import util.Recipe;

import java.util.Collections;

/**
 * Vendor Shipments API Recipe: Submit Shipments
 * ==============================================
 * 
 * This recipe demonstrates how to submit shipment requests for vendor orders.
 * Use this to create, update, or cancel shipments.
 * 
 * Use cases:
 * - Create new shipment requests
 * - Update existing shipment information
 * - Cancel shipments
 * 
 * API Operations used:
 * - submitShipments: Submit one or more shipment requests (async - returns transactionId)
 */
public class SubmitShipmentsRecipe extends Recipe {

    private static final Logger logger = LoggerFactory.getLogger(SubmitShipmentsRecipe.class);

    private final VendorShippingApi vendorShippingApi = new VendorShippingApi.Builder()
            .lwaAuthorizationCredentials(lwaCredentials)
            .endpoint(Constants.BACKEND_URL)
            .build();

    @Override
    public void start() {
        logger.info("=== Submitting Shipment Request ===");
        
        // Build a sample shipment request
        Shipment shipment = buildSampleShipment();
        
        // Submit the shipment
        String transactionId = submitShipment(shipment);
        
        if (transactionId != null) {
            logger.info("Shipment submitted successfully");
            logger.info("Transaction ID: {}", transactionId);
            logger.info("Use GetTransaction API to check processing status");
        }
    }

    /**
     * Build a sample shipment request.
     */
    private Shipment buildSampleShipment() {
        // Build purchase order items
        PurchaseOrderItems poItem = new PurchaseOrderItems()
                .itemSequenceNumber("001")
                .vendorProductIdentifier("9782700001659")
                .shippedQuantity(new ItemQuantity()
                        .amount(100)
                        .unitOfMeasure(ItemQuantity.UnitOfMeasureEnum.EACHES));

        PurchaseOrders purchaseOrder = new PurchaseOrders()
                .purchaseOrderNumber("1BBBAAAA")
                .items(Collections.singletonList(poItem));

        // Build shipment measurements
        TransportShipmentMeasurements measurements = new TransportShipmentMeasurements()
                .totalCartonCount(30)
                .totalPalletStackable(30)
                .totalPalletNonStackable(30);

        // Build the shipment
        return new Shipment()
                .vendorShipmentIdentifier("00050003")
                .buyerReferenceNumber("1234567")
                .transactionType(Shipment.TransactionTypeEnum.NEW)
                .transactionDate(OffsetDateTime.now())
                .shipmentFreightTerm(Shipment.ShipmentFreightTermEnum.COLLECT)
                .sellingParty(new PartyIdentification().partyId("PQRSS"))
                .shipFromParty(new PartyIdentification()
                        .partyId("999US")
                        .address(new Address()
                                .name("ABC electronics warehouse")
                                .addressLine1("DEF 1st street")
                                .city("Lisses")
                                .stateOrRegion("abcland")
                                .postalCode("91090")
                                .countryCode("DE")))
                .shipToParty(new PartyIdentification().partyId("ABCDF"))
                .shipmentMeasurements(measurements)
                .purchaseOrders(Collections.singletonList(purchaseOrder));
    }

    /**
     * Submit a shipment request.
     * 
     * @param shipment The shipment to submit
     * @return Transaction ID for tracking, or null if failed
     */
    public String submitShipment(Shipment shipment) {
        try {
            logger.info("Submitting shipment: {}", shipment.getVendorShipmentIdentifier());
            
            SubmitShipments request = new SubmitShipments()
                    .shipments(Collections.singletonList(shipment));

            SubmitShipmentConfirmationsResponse response = vendorShippingApi.submitShipments(request);

            if (response.getPayload() != null && response.getPayload().getTransactionId() != null) {
                return response.getPayload().getTransactionId();
            }
            
            logger.info("Shipment submitted (no transaction ID returned)");
            return null;

        } catch (ApiException e) {
            logger.error("API Error submitting shipment: {}", e.getMessage());
            throw new RuntimeException("Failed to submit shipment", e);
        } catch (LWAException e) {
            logger.error("Authentication error: {}", e.getMessage());
            throw new RuntimeException("Authentication error", e);
        }
    }
}

File: code-recipes/python/src/recipes/vendor_retail_procurement/shipments/submit_shipments_recipe.py
"""
Vendor Shipments API: Submit Shipments
=======================================

Submits shipment requests for vendor orders. Use to create, update, or cancel shipments.

Use cases:
- Create new shipment requests
- Update existing shipment information
- Cancel shipments

API operation: submitShipments
"""

import logging
from datetime import datetime, timezone
from typing import Any, Optional

from spapi import SPAPIClient, SPAPIConfig
from spapi.api.vendor_shipments_v1.vendor_shipment_api import VendorShipmentApi

from src import config
from src.util.recipe import Recipe

logger = logging.getLogger(__name__)


class SubmitShipmentsRecipe(Recipe):

    def __init__(
        self,
        sp_config: Optional[SPAPIConfig] = None,
        vendor_shipping_api: Optional[VendorShipmentApi] = None,
    ) -> None:
        super().__init__(config=sp_config)
        self._vendor_shipping_api = vendor_shipping_api

    @property
    def vendor_shipping_api(self) -> VendorShipmentApi:
        if self._vendor_shipping_api is None:
            client = SPAPIClient(
                self.config,
                oauth_endpoint=f"{config.backend_url}/auth/o2/token",
                endpoint=config.backend_url,
            )
            self._vendor_shipping_api = VendorShipmentApi(client.api_client)
            logger.info("VendorShipmentApi client initialized")
        return self._vendor_shipping_api

    def submit_shipment(
        self,
        vendor_shipment_id: str,
        buyer_reference_number: str,
        selling_party_id: str,
        ship_from_party_id: str,
        ship_to_party_id: str,
        purchase_orders: list,
    ) -> Optional[str]:
        """Submit a shipment request."""
        logger.info("Submitting shipment: %s", vendor_shipment_id)
        body = {
            "shipments": [
                {
                    "vendorShipmentIdentifier": vendor_shipment_id,
                    "buyerReferenceNumber": buyer_reference_number,
                    "transactionType": "New",
                    "transactionDate": datetime.now(timezone.utc).isoformat(),
                    "shipmentFreightTerm": "Collect",
                    "sellingParty": {"partyId": selling_party_id},
                    "shipFromParty": {
                        "partyId": ship_from_party_id,
                        "address": {
                            "name": "ABC electronics warehouse",
                            "addressLine1": "DEF 1st street",
                            "city": "Lisses",
                            "stateOrRegion": "abcland",
                            "postalCode": "91090",
                            "countryCode": "DE",
                        },
                    },
                    "shipToParty": {"partyId": ship_to_party_id},
                    "shipmentMeasurements": {
                        "totalCartonCount": 30,
                        "totalPalletStackable": 30,
                        "totalPalletNonStackable": 30,
                    },
                    "purchaseOrders": purchase_orders,
                }
            ]
        }
        response = self.vendor_shipping_api.submit_shipments(body=body)
        transaction_id = self._extract_transaction_id(response)
        if transaction_id:
            logger.info("Shipment submitted successfully")
            logger.info("Transaction ID: %s", transaction_id)
        return transaction_id

    def start(self) -> None:
        purchase_orders = [
            {
                "purchaseOrderNumber": "1BBBAAAA",
                "items": [
                    {
                        "itemSequenceNumber": "001",
                        "vendorProductIdentifier": "9782700001659",
                        "shippedQuantity": {"amount": 100, "unitOfMeasure": "Eaches"},
                    }
                ],
            }
        ]
        transaction_id = self.submit_shipment(
            vendor_shipment_id="00050003",
            buyer_reference_number="1234567",
            selling_party_id="PQRSS",
            ship_from_party_id="999US",
            ship_to_party_id="ABCDF",
            purchase_orders=purchase_orders,
        )
        logger.info("Shipment submitted. Transaction ID: %s", transaction_id)

    @staticmethod
    def _extract_transaction_id(response: Any) -> Optional[str]:
        if hasattr(response, "to_dict"):
            response = response.to_dict()
        payload = response.get("payload") or {}
        return payload.get("transaction_id")


if __name__ == "__main__":
    logging.basicConfig(level=logging.INFO)
    recipe = SubmitShipmentsRecipe()
    recipe.start()
